panic syncing new node

It starts of good until somewhere in the 6/16 phase it panics.

image: hermeznetwork/cdk-erigon:v2.61.2

```
[01-14|17:54:58.627] 
[1/16 L1Syncer] L1 Blocks processed progress (amounts): 80000/4727578 (1%)
[01-14|18:04:51.588] 
[2/16 L1InfoTree] L1 Blocks processed progress (amounts): 860000/4727578 (18%)
[01-14|18:07:47.047] 
[2/16 L1InfoTree] Processed 1250/52626 logs, 2% complete
[01-14|18:17:19.874] 
[3/16 Batches] Started downloading L2Blocks routine ID: 71386      
[01-14|18:17:26.025] 
[3/16 Batches] Saving stage progress     lastBlockHeight=100000
[01-14|18:43:36.209] 
[4/16 BlockHashes] Nothing new to transform
[01-14|18:44:06.218] 
[5/16 Senders] Recovery                  block_number=1730404 ch=9445/10000
[01-14|18:44:27.070] 
[5/16 Senders] Flushed buffer file       name=erigon-sortable-buf-1867699967
Committed State…
[01-14|18:48:36.753] 
[6/16 Execution] Executed blocks      number=64399 %=0.336
```
after some time
```
INFO] [01-14|19:36:06.753] [6/16 Execution] Executed blocks         number=1507042 %=7.867 blk/s=476.7 tx/s=476.7 Mgas/s=715.0 gasState%=+Inf batch=0B alloc=455.6MB sys=1.0GB
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x992bf7]

goroutine 118521 [running]:
github.com/erigontech/mdbx-go/mdbx.(*Txn).Info.func1(0xc0014d5590?, 0xc004332640?, 0x0?)
        github.com/erigontech/mdbx-go@v0.27.24/mdbx/txn.go:427 +0x17
github.com/erigontech/mdbx-go/mdbx.(*Txn).Info(0xc0042c4f30?, 0x0?)
        github.com/erigontech/mdbx-go@v0.27.24/mdbx/txn.go:427 +0x3e
github.com/ledgerwatch/erigon-lib/kv/mdbx.(*MdbxTx).CollectMetrics(0xc024732300)
        github.com/ledgerwatch/erigon-lib@v1.0.0/kv/mdbx/kv_mdbx.go:908 +0x32b
github.com/ledgerwatch/erigon/zk/utils.(*TxGasLogger).Start.func1()
        github.com/ledgerwatch/erigon/zk/utils/tx_gas_logger.go:59 +0xf6
created by github.com/ledgerwatch/erigon/zk/utils.(*TxGasLogger).Start in goroutine 65
        github.com/ledgerwatch/erigon/zk/utils/tx_gas_logger.go:51 +0x4f
```
It will start over at 0% and this continues for a very long time until you're lucky.

`[INFO] [01-14|19:36:06.753] [6/16 Execution] Executed blocks         number=1507042 %=7.867 blk/s=476.7 tx/s=476.7 Mgas/s=715.0 gasState%=+Inf batch=0B alloc=455.6MB sys=1.0GB`
after restart
`[INFO] [01-14|19:42:42.798] [6/16 Execution] Executed blocks         number=1705679 %=1.126 blk/s=615.3 tx/s=615.3 Mgas/s=1043.8 gasState%=+Inf batch=0B alloc=112.2MB sys=136.7MB`

When it restarts after the panic this is also visible several times:
```
[INFO] [01-14|20:26:20.596] [downloader] recover headers from db     left=65536
mdbx_setup_dxb:16114 opening after an unclean shutdown, but boot-id(9badc7db014b77f7-3c9ff0f220004669) is MATCH: rollback NOT needed, steady-sync NEEDED
```

--- a/erigon-lib/kv/mdbx/kv_mdbx.go
+++ b/erigon-lib/kv/mdbx/kv_mdbx.go
@@ -879,6 +879,9 @@ func (tx *MdbxTx) IsRo() bool     { return tx.readOnly }
 func (tx *MdbxTx) ViewID() uint64 { return tx.tx.ID() }
 
 func (tx *MdbxTx) CollectMetrics() {
+	if tx == nil {
+		return
+	}
 	if tx.db.opts.label != kv.ChainDB {
 		return
 	}

--- a/eth/stagedsync/stage_execute_zkevm.go
+++ b/eth/stagedsync/stage_execute_zkevm.go
@@ -105,7 +105,7 @@ func SpawnExecuteBlocksStageZk(s *StageState, u Unwinder, tx kv.RwTx, toBlock ui
 
 	stateStream := !initialCycle && cfg.stateStream && to-s.BlockNumber < stateStreamLimit
 
-	logger := utils.NewTxGasLogger(logInterval, s.BlockNumber, total, gasState, s.LogPrefix(), &batch, tx, stages.SyncMetrics[stages.Execution])
+	logger := utils.NewTxGasLogger(ctx, logInterval, s.BlockNumber, total, gasState, s.LogPrefix(), &batch, tx, stages.SyncMetrics[stages.Execution])
 	logger.Start()
 	defer logger.Stop()
 

--- a/zk/utils/tx_gas_logger.go
+++ b/zk/utils/tx_gas_logger.go
@@ -49,19 +49,27 @@ func NewTxGasLogger(logInterval time.Duration, logBlock, total, gasLimit uint64,
 
 func (g *TxGasLogger) Start() {
 	go func() {
-		for range g.logEvery.C {
-			g.logProgress()
-			g.logTx = g.lastLogTx
-			g.logBlock = g.currentBlockNum
-			g.logTime = time.Now()
-			g.gas = 0
-			if g.tx != nil {
-				g.tx.CollectMetrics()
-			}
-			g.metric.SetUint64(g.logBlock)
+		for {
+			func() {
+				// if tx gets committed during metrics collection, ensure we recover and start logging again
+				defer func() {
+					recover()
+				}()
+				for range g.logEvery.C {
+					g.logProgress()
+					g.logTx = g.lastLogTx
+					g.logBlock = g.currentBlockNum
+					g.logTime = time.Now()
+					g.gas = 0
+					if g.tx != nil {
+						g.tx.CollectMetrics()
+					}
+					g.metric.SetUint64(g.logBlock)
+				}
+			}()
+			time.Sleep(time.Second)
 		}
 	}()
-
 }
 
 func (g *TxGasLogger) Stop() {

--- a/zk/utils/tx_gas_logger.go
+++ b/zk/utils/tx_gas_logger.go
@@ -1,6 +1,7 @@
 package utils
 
 import (
+	"context"
 	"fmt"
 	"runtime"
 	"time"
@@ -13,6 +14,7 @@ import (
 )
 
 type TxGasLogger struct {
+	ctx             context.Context
 	logEvery        *time.Ticker
 	initialBlock    uint64
 	logBlock        uint64
@@ -30,8 +32,9 @@ type TxGasLogger struct {
 	metric          metrics.Gauge
 }
 
-func NewTxGasLogger(logInterval time.Duration, logBlock, total, gasLimit uint64, logPrefix string, batch *kv.PendingMutations, tx kv.RwTx, metric metrics.Gauge) *TxGasLogger {
+func NewTxGasLogger(ctx context.Context, logInterval time.Duration, logBlock, total, gasLimit uint64, logPrefix string, batch *kv.PendingMutations, tx kv.RwTx, metric metrics.Gauge) *TxGasLogger {
 	return &TxGasLogger{
+		ctx:          ctx,
 		logEvery:     time.NewTicker(logInterval),
 		initialBlock: logBlock,
 		logBlock:     logBlock,
@@ -49,13 +52,20 @@ func NewTxGasLogger(logInterval time.Duration, logBlock, total, gasLimit uint64,
 
 func (g *TxGasLogger) Start() {
 	go func() {
+		defer g.logEvery.Stop()
 		for {
-			func() {
-				// if tx gets committed during metrics collection, ensure we recover and start logging again
-				defer func() {
-					recover()
-				}()
-				for range g.logEvery.C {
+			select {
+			case <-g.ctx.Done():
+				log.Info(fmt.Sprintf("[%s] Stopping TxGasLogger", g.logPrefix))
+				return
+			case <-g.logEvery.C:
+				func() {
+					// if tx gets committed during metrics collection, ensure we recover and start logging again
+					defer func() {
+						if r := recover(); r != nil {
+							log.Warn(fmt.Sprintf("[%s] Encountered a panic during TxGasLogger, recovering...", g.logPrefix), "error", r)
+						}
+					}()
 					g.logProgress()
 					g.logTx = g.lastLogTx
 					g.logBlock = g.currentBlockNum
@@ -65,9 +75,8 @@ func (g *TxGasLogger) Start() {
 						g.tx.CollectMetrics()
 					}
 					g.metric.SetUint64(g.logBlock)
-				}
-			}()
-			time.Sleep(time.Second)
+				}()
+			}
 		}
 	}()
 }

